// @tryfinch/finch-api/resources/sandbox/sandbox.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sandbox = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const CompanyAPI = tslib_1.__importStar(require("./company.js"));
const company_1 = require("./company.js");
const DirectoryAPI = tslib_1.__importStar(require("./directory.js"));
const directory_1 = require("./directory.js");
const EmploymentAPI = tslib_1.__importStar(require("./employment.js"));
const employment_1 = require("./employment.js");
const IndividualAPI = tslib_1.__importStar(require("./individual.js"));
const individual_1 = require("./individual.js");
const PaymentAPI = tslib_1.__importStar(require("./payment.js"));
const payment_1 = require("./payment.js");
const ConnectionsAPI = tslib_1.__importStar(require("./connections/connections.js"));
const connections_1 = require("./connections/connections.js");
const JobsAPI = tslib_1.__importStar(require("./jobs/jobs.js"));
const jobs_1 = require("./jobs/jobs.js");
class Sandbox extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.connections = new ConnectionsAPI.Connections(this._client);
        this.company = new CompanyAPI.Company(this._client);
        this.directory = new DirectoryAPI.Directory(this._client);
        this.individual = new IndividualAPI.Individual(this._client);
        this.employment = new EmploymentAPI.Employment(this._client);
        this.payment = new PaymentAPI.Payment(this._client);
        this.jobs = new JobsAPI.Jobs(this._client);
    }
}
exports.Sandbox = Sandbox;
Sandbox.Connections = connections_1.Connections;
Sandbox.Company = company_1.Company;
Sandbox.Directory = directory_1.Directory;
Sandbox.Individual = individual_1.Individual;
Sandbox.Employment = employment_1.Employment;
Sandbox.Payment = payment_1.Payment;
Sandbox.Jobs = jobs_1.Jobs;
