// @tryfinch/finch-api/resources/sandbox/sandbox.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as CompanyAPI from "./company.mjs";
import { Company } from "./company.mjs";
import * as DirectoryAPI from "./directory.mjs";
import { Directory } from "./directory.mjs";
import * as EmploymentAPI from "./employment.mjs";
import { Employment } from "./employment.mjs";
import * as IndividualAPI from "./individual.mjs";
import { Individual } from "./individual.mjs";
import * as PaymentAPI from "./payment.mjs";
import { Payment } from "./payment.mjs";
import * as ConnectionsAPI from "./connections/connections.mjs";
import { Connections } from "./connections/connections.mjs";
import * as JobsAPI from "./jobs/jobs.mjs";
import { Jobs } from "./jobs/jobs.mjs";
export class Sandbox extends APIResource {
    constructor() {
        super(...arguments);
        this.connections = new ConnectionsAPI.Connections(this._client);
        this.company = new CompanyAPI.Company(this._client);
        this.directory = new DirectoryAPI.Directory(this._client);
        this.individual = new IndividualAPI.Individual(this._client);
        this.employment = new EmploymentAPI.Employment(this._client);
        this.payment = new PaymentAPI.Payment(this._client);
        this.jobs = new JobsAPI.Jobs(this._client);
    }
}
Sandbox.Connections = Connections;
Sandbox.Company = Company;
Sandbox.Directory = Directory;
Sandbox.Individual = Individual;
Sandbox.Employment = Employment;
Sandbox.Payment = Payment;
Sandbox.Jobs = Jobs;
