// @tryfinch/finch-api/resources/hris/company/pay-statement-item/rules.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../../core/resource.mjs";
import { ResponsesPage } from "../../../../core/pagination.mjs";
import { path } from "../../../../internal/utils/path.mjs";
export class Rules extends APIResource {
    /**
     * Custom rules can be created to associate specific attributes to pay statement
     * items depending on the use case. For example, pay statement items that meet
     * certain conditions can be labeled as a pre-tax 401k. This metadata can be
     * retrieved where pay statement item information is available.
     *
     * @example
     * ```ts
     * const rule =
     *   await client.hris.company.payStatementItem.rules.create();
     * ```
     */
    create(params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.post('/employer/pay-statement-item/rule', {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Update a rule for a pay statement item.
     *
     * @example
     * ```ts
     * const rule =
     *   await client.hris.company.payStatementItem.rules.update(
     *     'rule_id',
     *   );
     * ```
     */
    update(ruleID, params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.put(path `/employer/pay-statement-item/rule/${ruleID}`, {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * List all rules of a connection account.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ruleListResponse of client.hris.company.payStatementItem.rules.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/employer/pay-statement-item/rule', (ResponsesPage), {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Delete a rule for a pay statement item.
     *
     * @example
     * ```ts
     * const rule =
     *   await client.hris.company.payStatementItem.rules.delete(
     *     'rule_id',
     *   );
     * ```
     */
    delete(ruleID, params = {}, options) {
        const { entity_ids } = params ?? {};
        return this._client.delete(path `/employer/pay-statement-item/rule/${ruleID}`, {
            query: { entity_ids },
            ...options,
            __security: { bearerAuth: true },
        });
    }
}
