// @tryfinch/finch-api/resources/hris/company/pay-statement-item/rules.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rules = void 0;
const resource_1 = require("../../../../core/resource.js");
const pagination_1 = require("../../../../core/pagination.js");
const path_1 = require("../../../../internal/utils/path.js");
class Rules extends resource_1.APIResource {
    /**
     * Custom rules can be created to associate specific attributes to pay statement
     * items depending on the use case. For example, pay statement items that meet
     * certain conditions can be labeled as a pre-tax 401k. This metadata can be
     * retrieved where pay statement item information is available.
     *
     * @example
     * ```ts
     * const rule =
     *   await client.hris.company.payStatementItem.rules.create();
     * ```
     */
    create(params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.post('/employer/pay-statement-item/rule', {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Update a rule for a pay statement item.
     *
     * @example
     * ```ts
     * const rule =
     *   await client.hris.company.payStatementItem.rules.update(
     *     'rule_id',
     *   );
     * ```
     */
    update(ruleID, params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.put((0, path_1.path) `/employer/pay-statement-item/rule/${ruleID}`, {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * List all rules of a connection account.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const ruleListResponse of client.hris.company.payStatementItem.rules.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/employer/pay-statement-item/rule', (pagination_1.ResponsesPage), {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Delete a rule for a pay statement item.
     *
     * @example
     * ```ts
     * const rule =
     *   await client.hris.company.payStatementItem.rules.delete(
     *     'rule_id',
     *   );
     * ```
     */
    delete(ruleID, params = {}, options) {
        const { entity_ids } = params ?? {};
        return this._client.delete((0, path_1.path) `/employer/pay-statement-item/rule/${ruleID}`, {
            query: { entity_ids },
            ...options,
            __security: { bearerAuth: true },
        });
    }
}
exports.Rules = Rules;
