// @tryfinch/finch-api/resources/hris/benefits/benefits.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../core/resource.mjs";
import * as IndividualsAPI from "./individuals.mjs";
import { Individuals, } from "./individuals.mjs";
import { SinglePage } from "../../../core/pagination.mjs";
import { path } from "../../../internal/utils/path.mjs";
export class Benefits extends APIResource {
    constructor() {
        super(...arguments);
        this.individuals = new IndividualsAPI.Individuals(this._client);
    }
    /**
     * Creates a new company-wide deduction or contribution. Please use the
     * `/providers` endpoint to view available types for each provider.
     *
     * @example
     * ```ts
     * const createCompanyBenefitsResponse =
     *   await client.hris.benefits.create();
     * ```
     */
    create(params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.post('/employer/benefits', {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Lists deductions and contributions information for a given item
     *
     * @example
     * ```ts
     * const companyBenefit = await client.hris.benefits.retrieve(
     *   'benefit_id',
     * );
     * ```
     */
    retrieve(benefitID, query = {}, options) {
        return this._client.get(path `/employer/benefits/${benefitID}`, {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Updates an existing company-wide deduction or contribution
     *
     * @example
     * ```ts
     * const updateCompanyBenefitResponse =
     *   await client.hris.benefits.update('benefit_id');
     * ```
     */
    update(benefitID, params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.post(path `/employer/benefits/${benefitID}`, {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * List all company-wide deductions and contributions.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const companyBenefit of client.hris.benefits.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/employer/benefits', (SinglePage), {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Get deductions metadata
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const supportedBenefit of client.hris.benefits.listSupportedBenefits()) {
     *   // ...
     * }
     * ```
     */
    listSupportedBenefits(query = {}, options) {
        return this._client.getAPIList('/employer/benefits/meta', (SinglePage), {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
}
Benefits.Individuals = Individuals;
