// @tryfinch/finch-api/resources/hris/benefits/benefits.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Benefits = void 0;
const tslib_1 = require("../../../internal/tslib.js");
const resource_1 = require("../../../core/resource.js");
const IndividualsAPI = tslib_1.__importStar(require("./individuals.js"));
const individuals_1 = require("./individuals.js");
const pagination_1 = require("../../../core/pagination.js");
const path_1 = require("../../../internal/utils/path.js");
class Benefits extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.individuals = new IndividualsAPI.Individuals(this._client);
    }
    /**
     * Creates a new company-wide deduction or contribution. Please use the
     * `/providers` endpoint to view available types for each provider.
     *
     * @example
     * ```ts
     * const createCompanyBenefitsResponse =
     *   await client.hris.benefits.create();
     * ```
     */
    create(params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.post('/employer/benefits', {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Lists deductions and contributions information for a given item
     *
     * @example
     * ```ts
     * const companyBenefit = await client.hris.benefits.retrieve(
     *   'benefit_id',
     * );
     * ```
     */
    retrieve(benefitID, query = {}, options) {
        return this._client.get((0, path_1.path) `/employer/benefits/${benefitID}`, {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Updates an existing company-wide deduction or contribution
     *
     * @example
     * ```ts
     * const updateCompanyBenefitResponse =
     *   await client.hris.benefits.update('benefit_id');
     * ```
     */
    update(benefitID, params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.post((0, path_1.path) `/employer/benefits/${benefitID}`, {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * List all company-wide deductions and contributions.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const companyBenefit of client.hris.benefits.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/employer/benefits', (pagination_1.SinglePage), {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Get deductions metadata
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const supportedBenefit of client.hris.benefits.listSupportedBenefits()) {
     *   // ...
     * }
     * ```
     */
    listSupportedBenefits(query = {}, options) {
        return this._client.getAPIList('/employer/benefits/meta', (pagination_1.SinglePage), {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
}
exports.Benefits = Benefits;
Benefits.Individuals = individuals_1.Individuals;
