// @tryfinch/finch-api/resources/hris/benefits/individuals.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Individuals = void 0;
const resource_1 = require("../../../core/resource.js");
const pagination_1 = require("../../../core/pagination.js");
const path_1 = require("../../../internal/utils/path.js");
class Individuals extends resource_1.APIResource {
    /**
     * Enroll an individual into a deduction or contribution. This is an overwrite
     * operation. If the employee is already enrolled, the enrollment amounts will be
     * adjusted. Making the same request multiple times will not create new
     * enrollments, but will continue to set the state of the existing enrollment.
     *
     * @example
     * ```ts
     * const enrolledIndividualBenefitResponse =
     *   await client.hris.benefits.individuals.enrollMany(
     *     'benefit_id',
     *   );
     * ```
     */
    enrollMany(benefitID, params = undefined, options) {
        const { entity_ids, individuals } = params ?? {};
        return this._client.post((0, path_1.path) `/employer/benefits/${benefitID}/individuals`, {
            query: { entity_ids },
            body: individuals,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Lists individuals currently enrolled in a given deduction.
     *
     * @example
     * ```ts
     * const response =
     *   await client.hris.benefits.individuals.enrolledIDs(
     *     'benefit_id',
     *   );
     * ```
     */
    enrolledIDs(benefitID, query = {}, options) {
        return this._client.get((0, path_1.path) `/employer/benefits/${benefitID}/enrolled`, {
            query,
            ...options,
            __security: { bearerAuth: true },
        });
    }
    /**
     * Get enrollment information for the given individuals.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const individualBenefit of client.hris.benefits.individuals.retrieveManyBenefits(
     *   'benefit_id',
     * )) {
     *   // ...
     * }
     * ```
     */
    retrieveManyBenefits(benefitID, query = {}, options) {
        return this._client.getAPIList((0, path_1.path) `/employer/benefits/${benefitID}/individuals`, (pagination_1.SinglePage), { query, ...options, __security: { bearerAuth: true } });
    }
    /**
     * Unenroll individuals from a deduction or contribution
     *
     * @example
     * ```ts
     * const unenrolledIndividualBenefitResponse =
     *   await client.hris.benefits.individuals.unenrollMany(
     *     'benefit_id',
     *   );
     * ```
     */
    unenrollMany(benefitID, params = {}, options) {
        const { entity_ids, ...body } = params ?? {};
        return this._client.delete((0, path_1.path) `/employer/benefits/${benefitID}/individuals`, {
            query: { entity_ids },
            body,
            ...options,
            __security: { bearerAuth: true },
        });
    }
}
exports.Individuals = Individuals;
