// @tryfinch/finch-api/core/pagination.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var _AbstractPage_client;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "../internal/tslib.mjs";
import { FinchError } from "./error.mjs";
import { defaultParseResponse } from "../internal/parse.mjs";
import { APIPromise } from "./api-promise.mjs";
import { maybeObj } from "../internal/utils/values.mjs";
export class AbstractPage {
    constructor(client, response, body, options) {
        _AbstractPage_client.set(this, void 0);
        __classPrivateFieldSet(this, _AbstractPage_client, client, "f");
        this.options = options;
        this.response = response;
        this.body = body;
    }
    hasNextPage() {
        const items = this.getPaginatedItems();
        if (!items.length)
            return false;
        return this.nextPageRequestOptions() != null;
    }
    async getNextPage() {
        const nextOptions = this.nextPageRequestOptions();
        if (!nextOptions) {
            throw new FinchError('No next page expected; please check `.hasNextPage()` before calling `.getNextPage()`.');
        }
        return await __classPrivateFieldGet(this, _AbstractPage_client, "f").requestAPIList(this.constructor, nextOptions);
    }
    async *iterPages() {
        let page = this;
        yield page;
        while (page.hasNextPage()) {
            page = await page.getNextPage();
            yield page;
        }
    }
    async *[(_AbstractPage_client = new WeakMap(), Symbol.asyncIterator)]() {
        for await (const page of this.iterPages()) {
            for (const item of page.getPaginatedItems()) {
                yield item;
            }
        }
    }
}
/**
 * This subclass of Promise will resolve to an instantiated Page once the request completes.
 *
 * It also implements AsyncIterable to allow auto-paginating iteration on an unawaited list call, eg:
 *
 *    for await (const item of client.items.list()) {
 *      console.log(item)
 *    }
 */
export class PagePromise extends APIPromise {
    constructor(client, request, Page) {
        super(client, request, async (client, props) => new Page(client, props.response, await defaultParseResponse(client, props), props.options));
    }
    /**
     * Allow auto-paginating iteration on an unawaited list call, eg:
     *
     *    for await (const item of client.items.list()) {
     *      console.log(item)
     *    }
     */
    async *[Symbol.asyncIterator]() {
        const page = await this;
        for await (const item of page) {
            yield item;
        }
    }
}
export class SinglePage extends AbstractPage {
    constructor(client, response, body, options) {
        super(client, response, body, options);
        this.items = body || [];
    }
    getPaginatedItems() {
        return this.items ?? [];
    }
    nextPageRequestOptions() {
        return null;
    }
}
export class ResponsesPage extends AbstractPage {
    constructor(client, response, body, options) {
        super(client, response, body, options);
        this.responses = body.responses || [];
    }
    getPaginatedItems() {
        return this.responses ?? [];
    }
    nextPageRequestOptions() {
        return null;
    }
}
export class IndividualsPage extends AbstractPage {
    constructor(client, response, body, options) {
        super(client, response, body, options);
        this.individuals = body.individuals || [];
        this.paging = body.paging;
    }
    getPaginatedItems() {
        return this.individuals ?? [];
    }
    nextPageRequestOptions() {
        const offset = this.paging.offset ?? 0;
        const length = this.getPaginatedItems().length;
        const currentCount = offset + length;
        const totalCount = this.paging.count;
        if (!totalCount) {
            return null;
        }
        if (currentCount < totalCount) {
            return {
                ...this.options,
                query: {
                    ...maybeObj(this.options.query),
                    offset: currentCount,
                },
            };
        }
        return null;
    }
}
export class Page extends AbstractPage {
    constructor(client, response, body, options) {
        super(client, response, body, options);
        this.data = body.data || [];
        this.paging = body.paging;
    }
    getPaginatedItems() {
        return this.data ?? [];
    }
    nextPageRequestOptions() {
        const offset = this.paging.offset ?? 0;
        const length = this.getPaginatedItems().length;
        const currentCount = offset + length;
        const totalCount = this.paging.count;
        if (!totalCount) {
            return null;
        }
        if (currentCount < totalCount) {
            return {
                ...this.options,
                query: {
                    ...maybeObj(this.options.query),
                    offset: currentCount,
                },
            };
        }
        return null;
    }
}
