// @tryfinch/finch-api/resources/hris/hris.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.HRIS = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const DirectoryAPI = tslib_1.__importStar(require("./directory.js"));
const directory_1 = require("./directory.js");
const DocumentsAPI = tslib_1.__importStar(require("./documents.js"));
const documents_1 = require("./documents.js");
const EmploymentsAPI = tslib_1.__importStar(require("./employments.js"));
const employments_1 = require("./employments.js");
const IndividualsAPI = tslib_1.__importStar(require("./individuals.js"));
const individuals_1 = require("./individuals.js");
const PayStatementsAPI = tslib_1.__importStar(require("./pay-statements.js"));
const pay_statements_1 = require("./pay-statements.js");
const PaymentsAPI = tslib_1.__importStar(require("./payments.js"));
const payments_1 = require("./payments.js");
const BenefitsAPI = tslib_1.__importStar(require("./benefits/benefits.js"));
const benefits_1 = require("./benefits/benefits.js");
const CompanyAPI = tslib_1.__importStar(require("./company/company.js"));
const company_1 = require("./company/company.js");
class HRIS extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.company = new CompanyAPI.CompanyResource(this._client);
        this.directory = new DirectoryAPI.Directory(this._client);
        this.individuals = new IndividualsAPI.Individuals(this._client);
        this.employments = new EmploymentsAPI.Employments(this._client);
        this.payments = new PaymentsAPI.Payments(this._client);
        this.payStatements = new PayStatementsAPI.PayStatements(this._client);
        this.documents = new DocumentsAPI.Documents(this._client);
        this.benefits = new BenefitsAPI.Benefits(this._client);
    }
}
exports.HRIS = HRIS;
HRIS.CompanyResource = company_1.CompanyResource;
HRIS.Directory = directory_1.Directory;
HRIS.Individuals = individuals_1.Individuals;
HRIS.Employments = employments_1.Employments;
HRIS.Payments = payments_1.Payments;
HRIS.PayStatements = pay_statements_1.PayStatements;
HRIS.Documents = documents_1.Documents;
HRIS.Benefits = benefits_1.Benefits;
