// @tryfinch/finch-api/resources/hris/hris.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as DirectoryAPI from "./directory.mjs";
import { Directory, } from "./directory.mjs";
import * as DocumentsAPI from "./documents.mjs";
import { Documents, } from "./documents.mjs";
import * as EmploymentsAPI from "./employments.mjs";
import { Employments, } from "./employments.mjs";
import * as IndividualsAPI from "./individuals.mjs";
import { Individuals, } from "./individuals.mjs";
import * as PayStatementsAPI from "./pay-statements.mjs";
import { PayStatements, } from "./pay-statements.mjs";
import * as PaymentsAPI from "./payments.mjs";
import { Payments } from "./payments.mjs";
import * as BenefitsAPI from "./benefits/benefits.mjs";
import { Benefits, } from "./benefits/benefits.mjs";
import * as CompanyAPI from "./company/company.mjs";
import { CompanyResource } from "./company/company.mjs";
export class HRIS extends APIResource {
    constructor() {
        super(...arguments);
        this.company = new CompanyAPI.CompanyResource(this._client);
        this.directory = new DirectoryAPI.Directory(this._client);
        this.individuals = new IndividualsAPI.Individuals(this._client);
        this.employments = new EmploymentsAPI.Employments(this._client);
        this.payments = new PaymentsAPI.Payments(this._client);
        this.payStatements = new PayStatementsAPI.PayStatements(this._client);
        this.documents = new DocumentsAPI.Documents(this._client);
        this.benefits = new BenefitsAPI.Benefits(this._client);
    }
}
HRIS.CompanyResource = CompanyResource;
HRIS.Directory = Directory;
HRIS.Individuals = Individuals;
HRIS.Employments = Employments;
HRIS.Payments = Payments;
HRIS.PayStatements = PayStatements;
HRIS.Documents = Documents;
HRIS.Benefits = Benefits;
