// @tryfinch/finch-api/resources/hris/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payments = exports.PayStatements = exports.Individuals = exports.HRIS = exports.Employments = exports.Documents = exports.Directory = exports.CompanyResource = exports.Benefits = void 0;
var index_1 = require("./benefits/index.js");
Object.defineProperty(exports, "Benefits", { enumerable: true, get: function () { return index_1.Benefits; } });
var index_2 = require("./company/index.js");
Object.defineProperty(exports, "CompanyResource", { enumerable: true, get: function () { return index_2.CompanyResource; } });
var directory_1 = require("./directory.js");
Object.defineProperty(exports, "Directory", { enumerable: true, get: function () { return directory_1.Directory; } });
var documents_1 = require("./documents.js");
Object.defineProperty(exports, "Documents", { enumerable: true, get: function () { return documents_1.Documents; } });
var employments_1 = require("./employments.js");
Object.defineProperty(exports, "Employments", { enumerable: true, get: function () { return employments_1.Employments; } });
var hris_1 = require("./hris.js");
Object.defineProperty(exports, "HRIS", { enumerable: true, get: function () { return hris_1.HRIS; } });
var individuals_1 = require("./individuals.js");
Object.defineProperty(exports, "Individuals", { enumerable: true, get: function () { return individuals_1.Individuals; } });
var pay_statements_1 = require("./pay-statements.js");
Object.defineProperty(exports, "PayStatements", { enumerable: true, get: function () { return pay_statements_1.PayStatements; } });
var payments_1 = require("./payments.js");
Object.defineProperty(exports, "Payments", { enumerable: true, get: function () { return payments_1.Payments; } });
