// @tryfinch/finch-api/resources/access-tokens.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessTokens = void 0;
const resource_1 = require("../core/resource.js");
class AccessTokens extends resource_1.APIResource {
    /**
     * Exchange the authorization code for an access token
     */
    create(body, options) {
        const clientID = body.client_id || this._client.clientID;
        if (!clientID)
            throw new Error('client_id must be provided as an argument or with the FINCH_CLIENT_ID environment variable');
        const clientSecret = body.client_secret || this._client.clientSecret;
        if (!clientSecret)
            throw new Error('client_secret must be provided as an argument or with the FINCH_CLIENT_SECRET environment variable');
        const bodyWithReplacements = {
            ...body,
            client_id: clientID,
            client_secret: clientSecret,
        };
        const headersWithReplacements = {
            ...options?.headers,
            authorization: null,
        };
        return this._client.post('/auth/token', {
            body: bodyWithReplacements,
            ...options,
            headers: headersWithReplacements,
            __security: {},
        });
    }
}
exports.AccessTokens = AccessTokens;
