// @tryfinch/finch-api/resources/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = exports.Sandbox = exports.RequestForwarding = exports.Providers = exports.Payroll = exports.Jobs = exports.HRIS = exports.Connect = exports.Account = exports.AccessTokens = void 0;
const tslib_1 = require("../internal/tslib.js");
tslib_1.__exportStar(require("./shared.js"), exports);
var access_tokens_1 = require("./access-tokens.js");
Object.defineProperty(exports, "AccessTokens", { enumerable: true, get: function () { return access_tokens_1.AccessTokens; } });
var account_1 = require("./account.js");
Object.defineProperty(exports, "Account", { enumerable: true, get: function () { return account_1.Account; } });
var connect_1 = require("./connect/connect.js");
Object.defineProperty(exports, "Connect", { enumerable: true, get: function () { return connect_1.Connect; } });
var hris_1 = require("./hris/hris.js");
Object.defineProperty(exports, "HRIS", { enumerable: true, get: function () { return hris_1.HRIS; } });
var jobs_1 = require("./jobs/jobs.js");
Object.defineProperty(exports, "Jobs", { enumerable: true, get: function () { return jobs_1.Jobs; } });
var payroll_1 = require("./payroll/payroll.js");
Object.defineProperty(exports, "Payroll", { enumerable: true, get: function () { return payroll_1.Payroll; } });
var providers_1 = require("./providers.js");
Object.defineProperty(exports, "Providers", { enumerable: true, get: function () { return providers_1.Providers; } });
var request_forwarding_1 = require("./request-forwarding.js");
Object.defineProperty(exports, "RequestForwarding", { enumerable: true, get: function () { return request_forwarding_1.RequestForwarding; } });
var sandbox_1 = require("./sandbox/sandbox.js");
Object.defineProperty(exports, "Sandbox", { enumerable: true, get: function () { return sandbox_1.Sandbox; } });
var webhooks_1 = require("./webhooks.js");
Object.defineProperty(exports, "Webhooks", { enumerable: true, get: function () { return webhooks_1.Webhooks; } });
